### PR TITLE
Upgrade to Ubuntu Focal 20.04 LTS

### PR DESCRIPTION
**Pull request by nfsec**
- Upgrade Docker images to Ubuntu Focal 20.04 LTS.
- Standardization of pip install command.
- apt command optimization.

### CODE REVIEW

File: extra/docker/develop/Dockerfile [hunk 1]
```diff
@@ -51,20 +50,17 @@ RUN sudo apt-get install -y \
 	socat \
 	sshpass \
 	vim \
-	zsh
-
+	zsh \
 # Misc useful things when developing
-RUN sudo apt-get install -y  \
 	curl \
```
Comment (Arusekk) [inline extra/docker/develop/Dockerfile:59]: Does this work with the comment between escaped newlines?

File: extra/docker/develop/Dockerfile [hunk 2]
```diff
@@ -18,20 +17,20 @@ ENV PATH="/home/pwntools/.local/bin:${PATH}"
...
 
 # Requirements for running the tests
-RUN python  -m pip install --upgrade --requirement pwntools/docs/requirements.txt \
- && python3 -m pip install --upgrade --requirement pwntools/docs/requirements.txt
+RUN python3 -m pip install --upgrade --requirement pwntools/docs/requirements.txt
 
 # Python niceties for debugging
-RUN python  -m pip install -U ipython ipdb \
- && python3 -m pip install -U ipython ipdb
+RUN python3 -m pip install -U ipython ipdb
```
Comment (Arusekk) [inline extra/docker/develop/Dockerfile]: While at it, could these `pip install`s be moved after the `apt install`s? It looks like we might install several packages twice. The dependencies should be in a single RUN call preferably, too.  We should also remove the apt cache after `apt install` to take advantage of Docker layer caching (here and in base).

**Review by Arusekk — CHANGES_REQUESTED** [review]

File: extra/docker/develop/Dockerfile (hunk 1 above)
Comment (nfsec) [reply extra/docker/develop/Dockerfile:59]: Yes.

File: extra/docker/develop/Dockerfile (hunk 2 above)
Comment (nfsec) [reply extra/docker/develop/Dockerfile]: OK

**Comment by peace-maker** [comment]
This removes Python 2 from the images which breaks the local testing setup in `travis/docker`. It'd be great to keep python2 available in the image as long as pwntools still supports it.

File: extra/docker/base/Dockerfile [hunk 3]
```diff
@@ -3,26 +3,23 @@
 # Based on Ubuntu
 ############################################################
 
-FROM ubuntu:bionic
+FROM ubuntu:focal
```
Comment (Arusekk) [inline extra/docker/base/Dockerfile]: We can go even further now.

```suggestion
FROM ubuntu:jammy
```

File: extra/docker/base/Dockerfile [hunk 4]
```diff
@@ -3,26 +3,23 @@
...
+        locales \
         build-essential \
         elfutils \
         git \
         libssl-dev \
         libffi-dev \
-        python \
-        python-pip \
-        python-dev \
         python3 \
```
Comment (Arusekk) [inline extra/docker/base/Dockerfile:26]: We still need to support python 2 for testing.
```suggestion
        python2.7 \
        python2.7-dev \
        python3 \
```

File: extra/docker/base/Dockerfile [hunk 5]
```diff
@@ -35,14 +32,12 @@ RUN apt-get update \
         binutils-powerpc64-linux-gnu \
         binutils-sparc64-linux-gnu \
         tmux \
-    && pip install --upgrade pip \
-    && python -m pip install --upgrade pwntools \
-    && pip3 install --upgrade pip \
+    && locale-gen en_US.UTF-8 \
+    && update-locale LANG=en_US.UTF-8 \
+    && python3 -m pip install --upgrade pip \
```
Comment (Arusekk) [inline extra/docker/base/Dockerfile:43]:
```suggestion
    && python2.7 -m ensurepip \
    && python2.7 -m pip install --upgrade pip \
    && python2.7 -m pip install --upgrade pwntools \
    && python3 -m pip install --upgrade pip \
```

File: extra/docker/beta/Dockerfile [hunk 6]
```diff
@@ -1,7 +1,6 @@
 FROM pwntools/pwntools:stable
 
 USER root
-RUN pip install --upgrade git+https://github.com/Gallopsled/pwntools@beta
-RUN pip3 install --upgrade git+https://github.com/Gallopsled/pwntools@beta
+RUN python3 -m pip install --upgrade git+https://github.com/Gallopsled/pwntools@beta
```
Comment (Arusekk) [inline extra/docker/beta/Dockerfile]:
```suggestion
RUN python2.7 -m pip install --upgrade git+https://github.com/Gallopsled/pwntools@beta \
    && python3 -m pip install --upgrade git+https://github.com/Gallopsled/pwntools@beta
```

File: extra/docker/dev/Dockerfile [hunk 7]
```diff
@@ -1,7 +1,6 @@
 FROM pwntools/pwntools:stable
 
 USER root
-RUN pip install --upgrade git+https://github.com/Gallopsled/pwntools@dev
-RUN pip3 install --upgrade git+https://github.com/Gallopsled/pwntools@dev
+RUN python3 -m pip install --upgrade git+https://github.com/Gallopsled/pwntools@dev
```
Comment (Arusekk) [inline extra/docker/dev/Dockerfile]:
```suggestion
RUN python2.7 -m pip install --upgrade git+https://github.com/Gallopsled/pwntools@dev \
    && python3 -m pip install --upgrade git+https://github.com/Gallopsled/pwntools@dev
```

File: extra/docker/develop/Dockerfile [hunk 8]
```diff
@@ -5,8 +5,7 @@ ENV HISTFILE=/home/pwntools/.history
 
 # Uninstall existing versions of pwntools
 USER root
-RUN python -m pip uninstall -q -y pwntools \
- && python3 -m pip uninstall -q -y pwntools
+RUN python3 -m pip uninstall -q -y pwntools
```
Comment (Arusekk) [inline extra/docker/develop/Dockerfile]:
```suggestion
RUN python2.7 -m pip uninstall -q -y pwntools \
    && python3 -m pip uninstall -q -y pwntools
```

File: extra/docker/develop/Dockerfile [hunk 9]
```diff
@@ -18,20 +17,20 @@ ENV PATH="/home/pwntools/.local/bin:${PATH}"
 
 # Install Pwntools to the home directory, make it an editable install
 RUN git clone https://github.com/Gallopsled/pwntools \
- && python  -m pip install --upgrade --editable pwntools \
  && python3 -m pip install --upgrade --editable pwntools \
```
Comment (Arusekk) [inline extra/docker/develop/Dockerfile:22]:
```suggestion
 && python2.7 -m pip install --upgrade --editable pwntools \
 && python3 -m pip install --upgrade --editable pwntools \
```

File: extra/docker/develop/Dockerfile [hunk 10]
```diff
@@ -18,20 +17,20 @@ ENV PATH="/home/pwntools/.local/bin:${PATH}"
...
  && PWNLIB_NOTERM=1 pwn version
 
 # Requirements for running the tests
-RUN python  -m pip install --upgrade --requirement pwntools/docs/requirements.txt \
- && python3 -m pip install --upgrade --requirement pwntools/docs/requirements.txt
+RUN python3 -m pip install --upgrade --requirement pwntools/docs/requirements.txt
```
Comment (Arusekk) [inline extra/docker/develop/Dockerfile]:
```suggestion
RUN python2.7 -m pip install --upgrade --requirement pwntools/docs/requirements.txt \
    && python3 -m pip install --upgrade --requirement pwntools/docs/requirements.txt
```

File: extra/docker/develop/Dockerfile (hunk 2 above)
Comment (Arusekk) [inline extra/docker/develop/Dockerfile]:
```suggestion
RUN python2.7 -m pip install -U ipython ipdb \
    && python3 -m pip install -U ipython ipdb
```

File: extra/docker/stable/Dockerfile [hunk 11]
```diff
@@ -1,7 +1,6 @@
 FROM pwntools/pwntools:base
 
 USER root
-RUN pip install --upgrade git+https://github.com/Gallopsled/pwntools@stable
-RUN pip3 install --upgrade git+https://github.com/Gallopsled/pwntools@stable
+RUN python3 -m pip install --upgrade git+https://github.com/Gallopsled/pwntools@stable
```
Comment (Arusekk) [inline extra/docker/stable/Dockerfile]:
```suggestion
RUN python2.7 -m pip install --upgrade git+https://github.com/Gallopsled/pwntools@stable \
    && python3 -m pip install --upgrade git+https://github.com/Gallopsled/pwntools@stable
```

**Review by Arusekk — CHANGES_REQUESTED** [review]

**Pull request merged**